File: 448-find-all-numbers-disappeared-in-an-array/448-find-all-numbers-disappeared-in-an-array.cpp
class Solution {
public:
    vector<int> findDisappearedNumbers(vector<int>& nums) {
        vector<int> result;
        int n = nums.size();
        
        // sort the array
        for( int i = 0; i < n; i++) {
           while (nums[i] != i+1) {
               int d = nums[i] - 1;
               if( nums[i] != nums[d])
                   swap(nums[i], nums[d]);
               else
                   break;
           } 
        }
        
        //look for missing nums
        for( int i = 0; i < n; i++) {
            if ( nums[i] != i+1 )
                result.push_back(i+1);
        }    
        
        return result;
    }
};

File: 448-find-all-numbers-disappeared-in-an-array/448-find-all-numbers-disappeared-in-an-array.py
class Solution(object):
    def findDisappearedNumbers(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        for i in range(len(nums)):
            while nums[i] != i + 1:
                d = nums[i] - 1
                if nums[i] != nums[d]:
                    nums[i], nums[d] = nums[d], nums[i]
                else:
                    break
                    
        result = []
        for i in range(len(nums)):
            if nums[i] != i + 1:
                result.append(i+1)
        
        return result
                